storage/SMAIndexSubBlock.hpp

```diff
@@ -0,0 +1,418 @@
+struct SMAEntry {
+  attribute_id attribute;
+  // The type of the attribute contained in this entry.
+  TypeID type;
+  // A reference to the minimum value of the attribute.
+  EntryReference min_entry;
+  // A reference to the maximum value of the attribute.
+  EntryReference max_entry;
+  // The sum of the attribute. This is only used in the case of numeric types.
+  TypedValue sum;
+};
+
+}  // namespace sma_internal
+
+
+/**
+ * @brief Small Materialized Aggregate SubBlock.
+ * @details Keeps account of several types of aggregate functions per Block.
+ *          Currently supports min, max, sum, and count.
+ *          Physical organization of the block is as follows:
+ *
+ *          [SMAHeader][SMAEntry x number of indexed attributes]
+ */
+class SMAIndexSubBlock : public IndexSubBlock {
+ public:
+  SMAIndexSubBlock(const TupleStorageSubBlock &tuple_store,
+                   const IndexSubBlockDescription &description,
+                   const bool new_block,
+                   void *sub_block_memory,
+                   const std::size_t sub_block_memory_size);
+
+  /**
+   * @brief Frees data associated with variable length attributes.
+   * 
+   * Several of the data structure in this index are in heap, therefore it is
+   * important that the destructor be called when evicted. The variables which
+   * are held out of line are the variable length TypedValues and the comparators. 
+   */
+  ~SMAIndexSubBlock();
+
+  /**
+   * @brief Determine whether an IndexSubBlockDescription is valid for this
+   *        type of IndexSubBlock.
+   *
+   * @param relation The relation an index described by description would
+   *        belong to.
+   * @param description A description of the parameters for this type of
+   *        IndexSubBlock, which will be checked for validity.
+   * @return Whether description is well-formed and valid for this type of
+   *         IndexSubBlock belonging to relation (i.e. whether an IndexSubBlock
+   *         of this type, belonging to relation, can be constructed according
+   *         to description).
+   **/
+  static bool DescriptionIsValid(const CatalogRelationSchema &relation,
+                                 const IndexSubBlockDescription &description);
+
+  /**
+   * @brief Estimate the average number of bytes (including any applicable
+   *        overhead) used to index a single tuple in this type of
+   *        IndexSubBlock. Used by StorageBlockLayout::finalize() to divide
+   *        block memory amongst sub-blocks.
+   * @warning description must be valid. DescriptionIsValid() should be called
+   *          first if necessary.
+   *
+   * @param relation The relation tuples belong to.
+   * @param description A description of the parameters for this type of
+   *        IndexSubBlock.
+   *
+   * @return The average/amortized number of bytes used to index a single
+   *         tuple of relation in an IndexSubBlock of this type described by
+   *         description.
+   **/
```

Comment:
I noticed that the TODO to change our reliance on EstimateBytesPerTuple is present in the body of this function. A more appropriate place would be to add it either here or in the base class where this method is declared.